Common/Securities/ContractSecurityFilterUniverse.cs

```diff
@@ -0,0 +1,295 @@
+﻿/*
+ * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
+ * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Python.Runtime;
+using QuantConnect.Data;
+
+namespace QuantConnect.Securities
+{
+    /// <summary>
+    /// Base class for contract symbols filtering universes.
+    /// Used by OptionFilterUniverse and FutureFilterUniverse
+    /// </summary>
+    public abstract class ContractSecurityFilterUniverse<T> : IDerivativeSecurityFilterUniverse
+    where T: ContractSecurityFilterUniverse<T>
+    {
+        /// <summary>
+        /// Defines listed contract types
+        /// </summary>
+        public enum Type : int
+        {
+            /// <summary>
+            /// Standard contracts
+            /// </summary>
+            Standard = 1,
+
+            /// <summary>
+            /// Non standard weekly contracts
+            /// </summary>
+            Weeklys = 2
+        }
+
+        internal IEnumerable<Symbol> _allSymbols;
+
+        /// <summary>
+        /// The underlying price data
+        /// </summary>
+        public BaseData Underlying
+        {
+            get
+            {
+                // underlying value changes over time, so accessing it makes universe dynamic
+                _isDynamic = true;
+                return _underlying;
+            }
+        }
+
+        protected BaseData _underlying;
+
+        /// <summary>
+        /// True if the universe is dynamic and filter needs to be reapplied
+        /// </summary>
+        public bool IsDynamic => _isDynamic;
+
+        //TODO: Naming Conventions
+        internal bool _isDynamic;
+        protected Type _type = Type.Standard;
```

Comment:
Let's address the `TODO:` 😛 . Suggest we also rename `_type`. We might add a `remarks` to the `internal` type fields noting they are internal so that the extension methods can access them